python/ray/serve/tests/test_controller_recovery.py

```diff
@@ -249,5 +252,53 @@ def get_actor_info(name: str):
     assert actor_tag == get_actor_info(f"app_{V1.name}")[0]
 
 
+def test_replica_deletion_after_controller_recover(serve_instance):
+    """Test that replicas are deleted when controller is recovered"""
+
+    signal = SignalActor.remote()
+    controller = serve.context._global_client._controller
+
+    @serve.deployment
+    class V1:
+        async def __call__(self):
+            await signal.wait.remote()
+            return f"1|{os.getpid()}"
+
+    handle = serve.run(V1.bind(), name="app")
+    _ = handle.remote()
+    serve.delete("app", _blocking=False)
+
+    def check_replica(replica_state=None):
+        try:
+            replicas = ray.get(
+                controller._dump_replica_states_for_testing.remote("app_V1")
+            )
+        except RayTaskError as ex:
+            # Deployment is not existed any more.
+            if isinstance(ex, KeyError):
+                return []
+            # Unexpected exception raised.
+            raise ex
+        if replica_state is None:
+            replica_state = list(ReplicaState)
+        else:
+            replica_state = [replica_state]
+        return replicas.get(replica_state)
+
+    # Make sure the replica is in STOPPING state.
+    wait_for_condition(lambda: len(check_replica(ReplicaState.STOPPING)) > 0)
+    ray.kill(serve.context._global_client._controller, no_restart=False)
+    # Make sure the replica is in STOPPING state.
+    wait_for_condition(lambda: len(check_replica(ReplicaState.STOPPING)) > 0)
```

Comment:
Should we check that the replica state is stopping for a few extra iterations to be extra safe? To remove the possibility that the replica state is only briefly set to stopping after the controller recovers, and make sure it's actually stuck there until we unblock the signal.